### PR TITLE
ci: add automatic PR labeler

### PR DESCRIPTION
**Pull request by shmpwk**
Signed-off-by: Shumpei Wakabayashi <shumpei.wakabayashi@tier4.jp>

## Description

This PR labels opening PR automatically according to `label-path-mapping.yaml` once a commit is pushed.
Labels helps to visualize what category the PR belongs to.

## Related links

https://github.com/actions/labeler

## Tests performed

I tested this in my private environment.

## Notes for reviewers
- The target files are almost all in this repository except `LICENSE`, `NOTICE`, `*.repos`
- Once a PR is labeled, Labels won't disappear automatically even when the target files disappear at the future commit.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: .github/workflows/pr-labeler.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+name: pr-labeler
+on: [pull_request]
```
Comment (kenji-miyake) [inline .github/workflows/pr-labeler.yaml:2]: To be consistent with other workflow files.
```suggestion
on
  - pull_request:
```

File: label-path-mapping.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  - "*.json"
+  - "*.yaml"
+  - "*.cfg"
+  - .clang-format
+  - .gitignore
+  - .prettierignore
+document:
+  - docs/**/*
+  - "**/*.txt"
+  - "**/*.md"
```
Comment (kenji-miyake) [inline label-path-mapping.yaml:12]: Also .jpg, .png, .svg?

**Review by kenji-miyake — COMMENTED** [review]
Good idea, thanks! Added small comments.

**Comment by kenji-miyake** [comment]
@xmfcx Please comment if you have any thoughts.

File: .github/workflows/pr-labeler.yaml (hunk 1 above)
Comment (shmpwk) [reply .github/workflows/pr-labeler.yaml]: Ok, reflected it by https://github.com/autowarefoundation/autoware.universe/pull/2266/commits/2c00377ee913fd7c3d13fdd40e892493116492db

File: label-path-mapping.yaml (hunk 2 above)
Comment (shmpwk) [reply label-path-mapping.yaml:12]: Nice idea, reflected by https://github.com/autowarefoundation/autoware.universe/pull/2266/commits/2c00377ee913fd7c3d13fdd40e892493116492db

**Comment by shmpwk** [comment]
Thank you for fixing title.

File: .github/workflows/pr-labeler.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+name: pr-labeler
+on:
+  - pull_request
```
Comment (kenji-miyake) [inline .github/workflows/pr-labeler.yaml:3]: I think you need `:`.
```suggestion
  - pull_request:
```

Comment (kenji-miyake) [inline .github/workflows/pr-labeler.yaml:3]: Have you confirmed whether it works with PRs from forks? They have different permissions.

File: .github/workflows/pr-labeler.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+jobs:
+  label:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/labeler@v2
+        with:
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+          configuration-path: label-path-mapping.yaml
```
Comment (kenji-miyake) [inline .github/workflows/pr-labeler.yaml:12]: Let's put the file under `.github`.
Also, naming it `labeler.yml(yaml)` is recommended.
https://github.com/actions/labeler#create-githublabeleryml

```suggestion
          configuration-path: .github/labeler.yaml
```

File: .github/workflows/pr-labeler.yaml (hunk 3 above)
Comment (kenji-miyake) [reply .github/workflows/pr-labeler.yaml:3]: To understand the difference, please search by the word `pull_request_target`.

https://securitylab.github.com/research/github-actions-preventing-pwn-requests/
https://docs.github.com/ja/actions/using-workflows/events-that-trigger-workflows#pull_request_target

Comment (shmpwk) [reply .github/workflows/pr-labeler.yaml:3]: Not yet and I didn't notice that.
(In the first place, I should have create PR from the fork)
So it should follow like the other PR e.x.  https://github.com/autowarefoundation/autoware.universe/blob/main/.github/workflows/pre-commit.yaml#L8-L16 ?

Comment (kenji-miyake) [reply .github/workflows/pr-labeler.yaml:3]: `pre-commit.yaml` isn't related. These are the examples.
https://github.com/autowarefoundation/autoware.universe/blob/a2bcd7c531bffa5b7966f1c1b751a59733bfdcc6/.github/workflows/semantic-pull-request.yaml#L4
https://github.com/autowarefoundation/autoware.universe/blob/a2bcd7c531bffa5b7966f1c1b751a59733bfdcc6/.github/workflows/deploy-docs.yaml#L13

File: .github/workflows/pr-labeler.yaml (hunk 4 above)
Comment (shmpwk) [reply .github/workflows/pr-labeler.yaml]: Reflected by 
https://github.com/autowarefoundation/autoware.universe/pull/2266/commits/ce3bbd3f5b52de0cc1b7943448726ec499df539f

File: .github/workflows/pr-labeler.yaml (hunk 3 above)
Comment (shmpwk) [reply .github/workflows/pr-labeler.yaml]: This will be outdated

Comment (shmpwk) [reply .github/workflows/pr-labeler.yaml]: I finally understand it, thank you.
Reflected it by https://github.com/autowarefoundation/autoware.universe/pull/2266/commits/1a13aa2e91c91c50793285fa943cda75fd678784

File: .github/workflows/pr-labeler.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    types:
+      - opened
+      - edited
+      - synchronize
+
+jobs:
+  label:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/labeler@v2
```
Comment (kenji-miyake) [inline .github/workflows/pr-labeler.yaml:13]: v4 is the latest?
https://github.com/actions/labeler/releases/tag/v4

```suggestion
      - uses: actions/labeler@v4
```

Comment (shmpwk) [reply .github/workflows/pr-labeler.yaml]: Reflected by https://github.com/autowarefoundation/autoware.universe/pull/2266/commits/1a13aa2e91c91c50793285fa943cda75fd678784

**Review by kenji-miyake — APPROVED** [review]
I think it's okay to merge this and test it.

**Pull request merged**